proxy/context.go

```diff
@@ -0,0 +1,170 @@
+package proxy
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/zalando/skipper/routing"
+)
+
+type context struct {
+	responseWriter        http.ResponseWriter
+	request               *http.Request
+	response              *http.Response
+	route                 *routing.Route
+	deprecatedServed      bool
+	servedWithResponse    bool // to support the deprecated way independently
+	pathParams            map[string]string
+	stateBag              map[string]interface{}
+	originalRequest       *http.Request
+	originalResponse      *http.Response
+	outgoingHost          string
+	debugFilterPanics     []interface{}
+	outgoingDebugRequest  *http.Request
+	incomingDebugResponse *http.Response
+	startServe            time.Time
+}
+
+func defaultBody() io.ReadCloser {
+	return ioutil.NopCloser(&bytes.Buffer{})
+}
+
+func defaultResponse(r *http.Request) *http.Response {
+	return &http.Response{
+		StatusCode: http.StatusNotFound,
+		Header:     make(http.Header),
+		Body:       defaultBody(),
+		Request:    r,
+	}
+}
+
+func cloneURL(u *url.URL) *url.URL {
+	uc := *u
+	return &uc
+}
+
+func cloneRequestMetadata(r *http.Request) *http.Request {
```

Comment:
this is to provide information to the custom filters about the original incoming request before the filters could change it:

https://godoc.org/github.com/zalando/skipper/filters#FilterContext

```
    // The copy (deep) of the original incoming request or nil if the
    // implementation does not provide it.
    //
    // The object received from this method contains an empty body, and all
    // payload related properties have zero value.
```